src/hotspot/cpu/riscv/macroAssembler_riscv.hpp

```diff
@@ -376,8 +376,22 @@ class MacroAssembler: public Assembler {
     return ((predecessor & 0x3) << 2) | (successor & 0x3);
   }
 
+  void fence(uint32_t predecessor, uint32_t successor) {
+    if (UseZtso) {
+      if (pred_succ_to_membar_mask(predecessor, successor) & StoreLoad) {
+        // TSO allows for stores to be reordered after loads. When the compiler
+        // generates a fence to disallow that, we are required to generate the
+        // fence for correctness.
+      } else {
+        // TSO guarantees other fences already.
+        return;
+      }
+    }
+    Assembler::fence(predecessor, successor);
+  }
+
```

Comment:
```suggestion
  void fence(uint32_t predecessor, uint32_t successor) {
    if (UseZtso) {
      if (pred_succ_to_membar_mask(predecessor, successor) & StoreLoad) {
        // TSO allows for stores to be reordered after loads. When the compiler
        // generates a fence to disallow that, we are required to generate the
        // fence for correctness.
    Assembler::fence(predecessor, successor);
          } else {
        // TSO guarantees other orderings already.
      }
    }

  }

```

Reply:
@theRealAph this way we will miss fence completely when UseZtso is false

Reply:
I prefer:

if (UseZtso) {
  ....
  return;
}

<other code path>